### PR TITLE
Make sure dask array are writable via the __array__ protocol.

### LINKED ISSUE #6520: np.asarray returns non-writable array
**What happened**:
Behaviour change in the last release (2.23.0): converting a dask array to a numpy array results in a non-writable array when it has been created with `np.zeros_like` of a dask array.

**What you expected to happen**:
The returned numpy array should be consistently writable (or not?), regardless if the dask array have been created with `da.arange` or `da.zeros`, for example.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
import dask.array as da

dask_array = np.zeros_like(da.arange(20))
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```
Error:
```python
Traceback (most recent call last):

  File "<ipython-input-19-6eef1a4f1c7b>", line 5, in <module>
    numpy_array[0] = 1

ValueError: assignment destination is read-only
```

The following example works fine:
```python
dask_array = da.arange(20)
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```

Another example, which is not working with dask 0.23.0:
```python
dask_array = da.zeros((20))
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```

**Anything else we need to know?**:

Checking the numpy flags, it seems that the numpy array is not writable anymore.

**Environment**:

- Dask version: 2.23.0
- Python version: 3.7
- Operating System: Fedora
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by Carreau**
When trying to access dask array via the __array__ protocol, and those
are read only due to broadcast trick;  copy them to make them writable.

Closes #6520

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by Carreau** [comment]
If that's not the fix you like, would be good to at least keep the test.

**Comment by jakirkham** [comment]
Would suggest making this change in [`finalize`]( https://github.com/dask/dask/blob/a0dec10aef5cdac8b663ef29d7b228ecc4139a5a/dask/array/core.py#L980 ). That way this works for `__array__`, `Array.compute()`, etc.

**Comment by jsignell** [comment]
The conversation here seems to have stopped and I don't really understand the implications of @jakirkham's suggestion. Could you provide an example? Otherwise I am in favor of merging as-is.

**Comment by jakirkham** [comment]
Am just suggesting moving these two lines into [the `finalize` function]( https://github.com/dask/dask/blob/a0dec10aef5cdac8b663ef29d7b228ecc4139a5a/dask/array/core.py#L980 ). This is what [`__dask_postcompute__`]( https://github.com/dask/dask/blob/a0dec10aef5cdac8b663ef29d7b228ecc4139a5a/dask/array/core.py#L1099-L1100 ) calls, which is what follows things like `dask.compute`, `client.compute`, etc. Otherwise this only affects the case where the user calls `np.asarray(a)` (where `a` is a Dask Array).

**Comment by Carreau** [comment]
Sorry I'm just getting back to this, will do what @jakirkham suggested.

**Comment by Carreau** [comment]
updated.

File: dask/array/core.py [hunk 1]
```diff
@@ -989,6 +989,13 @@ def finalize(results):
     return unpack_singleton(results)
 
 
+def finalize(results):
+    res = _finalize(results)
+    if type(res) == np.ndarray and not res.flags.writeable:
```
Comment (Carreau) [inline dask/array/core.py:994]: I do not use `isinstance()` or otherwise it triggers on masked-arrays.

File: dask/array/core.py [hunk 2]
```diff
@@ -989,6 +989,13 @@ def finalize(results):
...
+        res = np.array(res, copy=True)
```
Comment (jakirkham) [inline dask/array/core.py:995]: Would suggest using the `.copy()` method so this works with `sparse` and `cupy` arrays as well.

```suggestion
        res = res.copy()
```

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py:994]: Maybe the suggestion below addresses the masked array case?

**Comment by shoyer** [comment]
As mentioned in #6520, I don't think this behavior is a good idea because it imposes unnecessary copies.

I also don't understand why the PR excludes writeable arrays. What is the use case for obtaining "always writeable arrays", when writing to such arrays may be unsafe?  Note that dask.array currently makes no guarantees about copying inputs, e.g.,
```
>>> import numpy as np
>>> import dask.array as da
>>> x = np.zeros(5)
>>> y = da.array(x)
>>> z = y.transpose().compute()
>>> z[0] = 1
>>> x
array([1., 0., 0., 0., 0.])
```

If we were going to make a breaking change here, I would be tempted to make computed dask arrays _never_ writeable. That would be safer than encouraging users to rely on dask's undocumented view/copy semantics.

**Comment by Carreau** [comment]
> I also don't understand why the PR excludes writeable arrays.

The goal is to only fix the problem introduced by #6491 which make zeros/ones/full read-only and breaks user code, I used `if type(res) == np.ndarray and not res.flags.writeable:` as the best guess to fix only the zeros/ones/full case, but I'm happy to use another condition.

I'm happy to go any route the devs here tink is the best.

File: dask/array/core.py (hunk 2 above)
Comment (Carreau) [reply dask/array/core.py:995]: It's up to you what you think we should do, so I tried to restrict myself to only copy the case introduces by zeros/ones/full, so I don't see any reason to copy those. I've pushed [another branch](https://travis-ci.org/github/Carreau/dask/builds/719613905) to see if `array.copy()` pass the tests.

**Comment by shoyer** [comment]
> > I also don't understand why the PR excludes writeable arrays.
> 
> The goal is to only fix the problem introduced by #6491 which make zeros/ones/full read-only and breaks user code, I used `if type(res) == np.ndarray and not res.flags.writeable:` as the best guess to fix only the zeros/ones/full case, but I'm happy to use another condition.

My personal opinion is that the user in #6520 was relying upon an implementation detail rather than an intentional part of Dask's API. I don't think we should be beholden to support it.

File: dask/array/core.py (hunk 1 above)
Comment (Carreau) [reply dask/array/core.py:994]: I did not remove the `type(...)` to avoid what I think is unnecessary, but let me know.

**Comment by Carreau** [comment]
> My personal opinion is that the user in #6520 was relying upon an implementation detail rather than an intentional part of Dask's API. I don't think we should be beholden to support it.

Maybe, but I think it's better to plan for such changes and potentially have a deprecation period. making the array RO was an non-expected side-effect which I think should be fixed in minors revision until there is proper agreement.

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py:994]: Yeah that's fine. We can always revisit if needed.

**Comment by jakirkham** [comment]
Yeah this concern ( https://github.com/dask/dask/pull/6525#issuecomment-677084228 ) does give me pause. Maybe making Dask Array's underlying data at input read-only would address this?

**Pull request closed without merging**